framework/src/utils/InputParameterWarehouse.C

```diff
@@ -26,36 +29,112 @@ InputParameterWarehouse::~InputParameterWarehouse()
 }
 
 InputParameters &
-InputParameterWarehouse::addInputParameters(const std::string & name, InputParameters parameters, THREAD_ID tid /* =0 */)
+InputParameterWarehouse::addInputParameters(std::string name, InputParameters parameters, THREAD_ID tid /* =0 */)
 {
-  // Check that the Parameters do not already exist
-  if (_name_to_shared_pointer[tid].find(name) != _name_to_shared_pointer[tid].end())
-    mooseError("An object already exists with the name " << name);
+
+  // Create the storage object
+  InputParametersContainer container;
+  container.object = name;
 
   // Create the actual InputParameters object to store and reference from the objects
   MooseSharedPointer<InputParameters> ptr(new InputParameters(parameters));
+  container.parameters = ptr;
+
+  // The object name, w/o the object base prefix (this is returned by MooseObject::name())
+  ptr->addParam<std::string>("name", name, "The name of the object");
+
+  // Set the system name
+  container.system = ptr->get<std::string>("_moose_base");
+  std::string system_name = container.system + "::" + container.object;
+
+  // Input file syntax, if it exists
+  if (ptr->isParamValid("_syntax"))
+    container.syntax = ptr->get<std::string>("_syntax");
+
+  // User-defined tag, if it exists (optional)
+  if (ptr->isParamValid("control_tag"))
+    container.tag = ptr->get<std::string>("control_tag");
+
+  // Check that the Parameters do not already exist
+  if (_system_to_index[tid].find(system_name) != _system_to_index[tid].end())
+    mooseError("A " << container.system << " object already exists with the name " << name << ".\n");
 
   // Set the name and tid parameters
-  ptr->addParam<std::string>("name", name, "The complete name of the object");
   ptr->addPrivateParam<THREAD_ID>("_tid", tid);
   ptr->allowCopy(false);  // no more copies allowed
 
   // Store the object in the warehouse
-  _name_to_shared_pointer[tid].insert(std::pair<std::string, MooseSharedPointer<InputParameters> >(name, ptr));
+  _input_parameters[tid].push_back(container);
+  unsigned int index = _input_parameters.size() - 1;
+
+  // Store the name for checking that names are unique and name-based lookup
+  _system_to_index[tid].insert(std::pair<std::string, unsigned int>(system_name, index));
 
-  return *_name_to_shared_pointer[tid][name];
+  if (ptr->isParamValid("_syntax"))
+  {
+    std::string syntax_name = container.syntax + "/" + container.object;
+    _syntax_to_index[tid].insert(std::pair<std::string, unsigned int>(syntax_name, index));
+  }
+
+  if (ptr->isParamValid("control_tag"))
+  {
+    std::string tag_name = container.tag + "::" + container.object;
+    _tag_to_index[tid].insert(std::pair<std::string, unsigned int>(tag_name, index));
+  }
+
+  return *ptr;
 }
```

Comment:
I think I'm with @friedmud on this. I thought the design was that we could have any number of "names", "tags", "syntaxes", whatever you want to call them. They'd all be part of some set of tags for each object. If we collapse everything into one format, it should matter how they get there. The way we search through them and process them would be the same and remove all of the if/then/else clauses you have in methods like these.  The parser would start the `std::set` the user could append to it and so forth. This whole tag system should be a new object that handles all of these formats and just responds to requests. There's way too much parsing going on all over the place to work with strings.